### ISSUE TITLE
[flaky test]: [sig-node] Probing container should be ready immediately after startupProbe succeeds

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/kubernetes#99979]
#### Which jobs are flaking:

Seems to be impacting a few KIND jobs on the master branch the most, but also assorted other jobs

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=%20more%20than%205s%20after%20startupProbe%20succeeded

#### Which test(s) are flaking:

`[sig-node] Probing container should be ready immediately after startupProbe succeeds`

#### Testgrid link:

see the triage link above

#### Reason for failure:

`Pod became ready in 1m0.009381099s, more than 5s after startupProbe succeeded. It means that the delay readiness probes were not initiated immediately after startup finished.`

#### Anything else we need to know:

Reported by @ehashman https://kubernetes.slack.com/archives/C0BP8PW9G/p1615260313257400

Sample failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99375/pull-kubernetes-e2e-kind-ipv6/1369116272494645248

/sig node

**Comment by BenTheElder** [comment]
xref probing tests https://github.com/kubernetes/kubernetes/issues/99909#issuecomment-793422923

**Comment by ehashman** [comment]
/assign @matthyx 
ref https://github.com/kubernetes/kubernetes/pull/99998

/triage accepted

**Comment by matthyx** [comment]
Thanks Elana, still trying today...
Does anyone know if we can use the kubelet logs in e2e tests?

**Comment by BenTheElder** [comment]
> Does anyone know if we can use the kubelet logs in e2e tests?

Not very portably in cluster e2e tests at least. Maybe in node_e2e instead.

**Comment by manfredlift** [comment]
Have you considered that the issue is not with the test but with the underlying code in kubelet like detailed in this issue https://github.com/kubernetes/kubernetes/issues/95140?
 
 Like detailed in the issue I can reliably reproduce (on k8s 1.18) the startup probe passing within a few seconds, but the container is only marked as ready after about a minute.

**Comment by matthyx** [comment]
@manfredlift there is already a fix in the 1.21 release for that (part of a bigger PR):
https://github.com/kubernetes/kubernetes/pull/98376/commits/4870e64ac110d6bf62dbd270f8cef557c4bd6d53

**Comment by thejoycekung** [comment]
Some recent flakes:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1373982722380271616
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1373921315332820992
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1373905963559424000
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1373829207318073344
- https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1373737099076833280

**Comment by BenTheElder** [comment]
one of the most common flakes in kind jobs currently

http://storage.googleapis.com/k8s-metrics/flakes-latest.json

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=Probing%20container%20should%20be%20ready%20immediately%20after%20startupProbe%20succeeds

failures like:
> Mar 25 20:09:26.384: Pod became ready in 6.00715356s, more than 5s after startupProbe succeeded. It means that the delay readiness probes were not initiated immediately after startup finished.


>Mar 25 17:03:42.508: Pod became ready in 42.007286524s, more than 5s after startupProbe succeeded. It means that the delay readiness probes were not initiated immediately after startup finished.

**Comment by matthyx** [comment]
> one of the most common flakes in kind jobs currently

Maybe we could speed up the review on https://github.com/kubernetes/kubernetes/pull/99998

**Comment by BenTheElder** [comment]
Thank you I will take a look. I'm making a pass through flakes tonight trying to identify if anything is on kind's end 😅

**Comment by matthyx** [comment]
> Thank you I will take a look. I'm making a pass through flakes tonight trying to identify if anything is on kind's end

I don't think it's kind... since running on my PC with kind it works 100%, I think the issue is either the resources... or contention in the queue as mentioned to @SergeyKanzhelev in the comment https://github.com/kubernetes/kubernetes/pull/99998#issuecomment-801427213

**Comment by SergeyKanzhelev** [comment]
commented on PR with the suggested fix

**Comment by ehashman** [comment]
/reopen

https://github.com/kubernetes/kubernetes/pull/99998#issuecomment-829510526

**Comment by matthyx** [comment]
Yeah... 🥺

**Comment by matthyx** [comment]
Thanks for your great work in keeping things in check @ehashman

**Comment by ehashman** [comment]
FYI @SergeyKanzhelev

**Comment by matthyx** [comment]
I have (finally) understood what's happening, and I can even explain why the deflake made things worse...

Please have a look at the sleep before the worker starts to process probes:
https://github.com/kubernetes/kubernetes/blob/1eccb41fa878c5892b8278d234872b1b842448fe/pkg/kubelet/prober/worker.go#L132-L139

So, you have to remember that we have one worker per pod and per probe, and since we have these probe settings for the test:
https://github.com/kubernetes/kubernetes/blob/8daa2e8a4e705514282760a816108ca82f45f813/test/e2e/common/node/container_probe.go#L402-L411

We have `startupProbe` (PeriodSeconds=10) starting sometimes immediately (if the Kubelet has started since more than 10s) or after a quick sleep of rand*10s. On the contrary, `readinessProbe` will likely sleep most of the time for rand*60s.

This causes a delay in the processing of `manualTriggerCh` notifications:
https://github.com/kubernetes/kubernetes/blob/1eccb41fa878c5892b8278d234872b1b842448fe/pkg/kubelet/prober/worker.go#L156-L167

If we are lucky, this delay is less than 5s and the test passes:
https://github.com/kubernetes/kubernetes/blob/8daa2e8a4e705514282760a816108ca82f45f813/test/e2e/common/node/container_probe.go#L439-L441

Now you can understand why the deflaked settings were actually worse:
https://github.com/kubernetes/kubernetes/blob/bcd582030fdb1e3a5ac74c7b01a99fbfd9069a68/test/e2e/common/node/container_probe.go#L402-L413

**Comment by matthyx** [comment]
The reason I didn't spot these was that I usually start my `kindest` cluster once for all and iterate on the tests... so the kubelet had plenty of time to start and I never ended in that sleep.

**Comment by matthyx** [comment]
For the solution, I suggest we either remove the sleep completely, or we find something that would make all 3 probe workers start at the same time to avoid sync issues...

For the record, I was able to spot this issue thanks to the following log error (because the channel was overflowing):
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/prober/prober_manager.go#L251-L256

What do you think @ehashman @SergeyKanzhelev ?

**Comment by matthyx** [comment]
I have found another issue in that code:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/prober/prober_manager.go#L240-L260

If `readinessProbe` has already ran (and failed), we won't probe it via `manualTriggerCh` and therefore wait for the next tick, making the test fail.

**Comment by ehashman** [comment]
> We have startupProbe (PeriodSeconds=10) starting sometimes immediately (if the Kubelet has started since more than 10s) or after a quick sleep of rand10s. On the contrary, readinessProbe will likely sleep most of the time for rand60s.

This seems consistent with what I found when I was adding e2es for the new probe-level termination grace period and why I did my timing math the way I did.

**Comment by ehashman** [comment]
/reopen

I don't think this is fixed, I just hit it again and gubernator is still showing plenty of flakes: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=%20more%20than%205s%20after%20startupProbe%20succeeded

**Comment by matthyx** [comment]
@ehashman @BenTheElder how to check the exact commit used for each test?

For instance, I am not sure (from the name) that `ci-kubernetes-kind-e2e-parallel-1-21` should include my latest fix...

**Comment by matthyx** [comment]
In your PR the fix wasn't there:
https://github.com/ehashman/kubernetes/blob/init-container-coverage/pkg/kubelet/prober/prober_manager.go#L245

**Comment by pacoxu** [comment]
@matthyx you may send cherry-pick to 1.21 for `ci-kubernetes-kind-e2e-parallel-1-21`.

**Comment by pacoxu** [comment]
`ci-kubernetes-e2e-gce-stable1-beta-upgrade-cluster-new-parallel` is using [1.20](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-stable1-beta-upgrade-cluster-new-parallel/1401996899036172288/artifacts/bootstrap-e2e-minion-group-d4gz/kubelet.log).

**Comment by matthyx** [comment]
@ehashman I have found this one that failed **even with the fix**:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102697/pull-kubernetes-e2e-kind-ipv6/1402400393228455936
corresponding to:
https://github.com/kubernetes/kubernetes/pull/102697

**Comment by matthyx** [comment]
I don't know what could explain this atm... but it seems we still have delay propagating container statuses.

**Comment by BenTheElder** [comment]
> @ehashman @BenTheElder how to check the exact commit used for each test?

In the job artifacts there are multiple places with this info including the clone records/logs, and started.json

In testgrid it's a row.

**Comment by ehashman** [comment]
> @ehashman I have found this one that failed **even with the fix**:
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102697/pull-kubernetes-e2e-kind-ipv6/1402400393228455936
> corresponding to:
> #102697

yeah afaict since the tests ran on my PR after your fix was merged into master, it showed the flake was still present :( Prow will always use the latest master when running CI.

**Comment by matthyx** [comment]
> yeah afaict since the tests ran on my PR after your fix was merged into master, it showed the flake was still present :( Prow will always use the latest master when running CI.

Sorry for doubting what you said... I was probably in deny and hoped this annoyance was fixed once for all :(

**Comment by ehashman** [comment]
Plan: we'll see if Clayton's pod lifecycle refactor and/or the jitter refactor for probes affects this.

**Comment by matthyx** [comment]
> Plan: we'll see if Clayton's pod lifecycle refactor and/or the jitter refactor for probes affects this.

Clayton's PR #102344
jitter refactor #102064

**Comment by matthyx** [comment]
@ehashman since #102344 isn't (likely) coming to 1.22 we could cut off 90% of flakes by increasing to 15s delay:
https://github.com/kubernetes/kubernetes/blob/625597425b014d3a201babfa49f9dfe08e8c7437/test/e2e/common/node/container_probe.go#L448

**Comment by matthyx** [comment]
@ehashman it seems it fails even after merging #102344 so let's continue our discussion on slack

**Comment by SergeyKanzhelev** [comment]
this is not a regression, our suggestion is not block release on it. @kubernetes/release-managers feel free to remove the milestone

**Comment by BenTheElder** [comment]
> this is not a regression,

In Kubelet maybe, but in the flakiness of the e2e tests it was. Unfortunately we can't easily explore the current state :/

**Comment by BenTheElder** [comment]
In the release blocking jobs the frequency is not horrible now, so unblocking is probably fine, it is not as bad as it was.

Outside of the release cycle this is still the flakiest test in some jobs, see e.g. https://testgrid.k8s.io/sig-release-master-blocking#kind-master-parallel&width=5 where it's the only flake source we have on record currently.

**Comment by ehashman** [comment]
@BenTheElder if we end up needing it, we have a timing/period adjustment PR we can take that should help with the flakiness with https://github.com/kubernetes/kubernetes/pull/103593
